cleanrl/rnd_ppo.py

```diff
@@ -0,0 +1,1079 @@
+# if args.prod_mode:
+#     envs = VecPyTorch(
+#         SubprocVecEnv([make_env(args.gym_id, args.seed+i, i) for i in range(N)], "fork"),
+#         device
+#     )
+assert isinstance(envs.action_space, Discrete), "only discrete action space is supported"
+
+# ALGO LOGIC: initialize agent here:
+class Scale(nn.Module):
+    def __init__(self, scale):
+        super().__init__()
+        self.scale = scale
+
+    def forward(self, x):
+        return x * self.scale
+
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+class Agent(nn.Module):
+    def __init__(self, frames=4):
+        super(Agent, self).__init__()
+        self.network = nn.Sequential(
+            Scale(1/255),
+            layer_init(nn.Conv2d(frames, 32, 8, stride=4)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(32, 64, 4, stride=2)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(64, 64, 3, stride=1)),
+            nn.ReLU(),
+            nn.Flatten(),
+            layer_init(nn.Linear(3136, 256)),
+            nn.ReLU(),
+            layer_init(nn.Linear(
+                256,
+                448)),
+            nn.ReLU()
+        )
+        self.extra_layer = nn.Sequential(
+            layer_init(nn.Linear(448, 448), std=0.1),
+            nn.ReLU()
+        )
+        self.actor = nn.Sequential(
+            layer_init(nn.Linear(448, 448), std=0.01), 
+            nn.ReLU(),  
+            layer_init(nn.Linear(448, envs.action_space.n), std=0.01))
+        self.critic_ext = layer_init(nn.Linear(448, 1), std=0.01)
+        self.critic_int = layer_init(nn.Linear(448, 1), std=0.01)
+
+    def forward(self, x):
+        return self.network(x)
+
+    def get_action(self, x, action=None):
+        logits = self.actor(self.forward(x))
+        probs = Categorical(logits=logits)
+        if action is None:
+            action = probs.sample()
+        return action, probs.log_prob(action), probs.entropy()
+
+    def get_value(self, x):
+        features = self.forward(x)
+        return self.critic_ext(self.extra_layer(features) + features), self.critic_int(self.extra_layer(features) + features)
+
+agent = Agent().to(device)
+
+rnd_model = RNDModel(4, envs.action_space.n).to(device)
+
+optimizer = optim.Adam(list(agent.parameters()) + list(rnd_model.predictor.parameters()), lr=args.learning_rate, eps=1e-5)
+if args.anneal_lr:
+    # https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/ppo2/defaults.py#L20
+    lr = lambda f: f * args.learning_rate
+
+reward_rms = RunningMeanStd()
+obs_rms = RunningMeanStd(shape=(1, 1, 84, 84))
+
+discounted_reward = RewardForwardFilter(args.int_gamma)
+
+
+# ALGO Logic: Storage for epoch data
+obs = torch.zeros((args.num_steps, args.num_envs) + envs.observation_space.shape).to(device)
+actions = torch.zeros((args.num_steps, args.num_envs)).to(device)
```

Comment:
Please consider changing it back to 
```python
actions = torch.zeros((args.num_steps, args.num_envs) + envs.action_space.shape).to(device)
```
Although in Atari `envs.action_space.shape=()`, this is important for other domains.